### ISSUE TITLE
Bug with MvxSpinner when used in an MvxItemList

### ISSUE THREAD

**Issue by JimWilcox3** [MvvmCross/MvvmCross#944]
I have written a self contained viewmodel with self creating data with a view and xml for the view and the item view.  This shows the bug where the MvxSpinner selected value is sometimes wrong for existing data.

The View is:

```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Android.App;
using Android.Content;
using Android.OS;
using Android.Runtime;
using Android.Views;
using Android.Widget;
using Cirrious.MvvmCross.Droid.Views;

namespace MPS_Mobile_Driver.Droid.Views
{
    [Activity(Label = "Test View")]
    public class TestView : MvxActivity
    {
        protected override void OnCreate(Bundle bundle)
        {
            base.OnCreate(bundle);
            SetContentView(Resource.Layout.TestView);

            // Create your application here
        }
    }
}
```
The View Model is:
```
using Cirrious.MvvmCross.ViewModels;
using MPS_Mobile_Driver.Droid.Common;
using MPS_Mobile_Driver.Droid.DataModel;
using MPS_Mobile_Shared.Log;
using MPSDataService;
using System;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Threading.Tasks;

namespace MPS_Mobile_Driver.Droid.ViewModels
{
    public class TestViewModel
      : MvxViewModel
    {
        public void Init(Guid ID)
        {
            ObservableCollection<TestItem> t = new ObservableCollection<TestItem>();
            TestItem i = new TestItem() { TagColor = "Blue", TagLot = "123", InventoryNumber = "1" };
            t.Add(i);
            i = new TestItem() { TagColor = "Blue", TagLot = "123", InventoryNumber = "2" };
            t.Add(i);
            i = new TestItem() { TagColor = "Blue", TagLot = "123", InventoryNumber = "3" };
            t.Add(i);
            i = new TestItem() { TagColor = "Blue", TagLot = "123", InventoryNumber = "4" };
            t.Add(i);
            i = new TestItem() { TagColor = "Blue", TagLot = "123", InventoryNumber = "5" };
            t.Add(i);
            i = new TestItem() { TagColor = "Blue", TagLot = "123", InventoryNumber = "6" };
            t.Add(i);
            i = new TestItem() { TagColor = "Blue", TagLot = "123", InventoryNumber = "7" };
            t.Add(i);

            Items = t;        
        }



        private ObservableCollection<TestItem> _Items;
        public ObservableCollection<TestItem > Items
        {
            get
            {
                return _Items;
            }
            set
            {
                _Items = value;
                RaisePropertyChanged(() => Items);
            }
        }

    }

    public class TestItem : INotifyPropertyChanged
    {
        public event PropertyChangedEventHandler PropertyChanged;

        protected void RaisePropertyChanged(string propertyName)
        {
            System.ComponentModel.PropertyChangedEventHandler propertyChanged = this.PropertyChanged;
            if ((propertyChanged != null))
            {
                propertyChanged(this, new System.ComponentModel.PropertyChangedEventArgs(propertyName));
            }
        }

        private string _TagColor;
        public string TagColor
        {
            get
            {
                return _TagColor;
            }
            set
            {
                _TagColor = value;
                this.RaisePropertyChanged("TagColor");
            }
        }

        private string _TagLot;
        public string TagLot
        {
            get
            {
                return _TagLot;
            }
            set
            {
                _TagLot = value;
                this.RaisePropertyChanged("TagLot");
            }
        }

        private string _InventoryNumber;
        public string InventoryNumber
        {
            get
            {
                return _InventoryNumber;
            }
            set
            {
                _InventoryNumber = value;
                this.RaisePropertyChanged("InventoryNumber");
            }
        }

        private static string[] _TagColors = { "Yellow", "Brown", "White", "Blue", "Orange", "Red", "Green", "Purple" };
        public string[] TagColors
        {
            get { return _TagColors; }
        }

    }
}
```
The xml for the View "TestView" is:
```
<?xml version="1.0" encoding="utf-8"?>
<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
              xmlns:local="http://schemas.android.com/apk/res-auto"
              android:orientation="vertical"
              android:layout_width="fill_parent"
              android:layout_height="fill_parent">

  <Mvx.MvxListView
    android:layout_width="fill_parent"
    android:layout_height="wrap_content"
    android:layout_marginTop="5dp"
    local:MvxBind="ItemsSource Items"
    local:MvxItemTemplate="@layout/testitemview" />

</LinearLayout>
```
The xml for the item view "TestItemView" is:
```
<?xml version="1.0" encoding="utf-8"?>
<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
              xmlns:local="http://schemas.android.com/apk/res-auto"
              android:orientation="horizontal"
              android:layout_width="fill_parent"
              android:layout_height="fill_parent">

  <MvxSpinner
    android:layout_width="130dp"
    android:layout_height="wrap_content"
    local:MvxBind="ItemsSource TagColors; SelectedItem TagColor"
    android:id="@+id/spinner1" />
  <EditText
    android:layout_width="130sp"
    android:layout_height="wrap_content"
    local:MvxBind="Text TagLot" />
  <EditText
    android:layout_width="150sp"
    android:layout_height="wrap_content"
    local:MvxBind="Text InventoryNumber" />

</LinearLayout>
```
When you plug this into an Android app using Xamarin and MvvmCross, the color for each line should be Blue.  On the first line it changes to Yellow which is the first item on the list.  If you change line number 6, it then it changes lines number 2 to Yellow.  Please let me know if there is a workaround for this or if you need more information.

Thanks,
Jim

**Comment by Cheesebaron** [comment]
Could you make a GitHub repo with this code?

**Comment by slodge** [comment]
This is just .net.... try overriding `Equals` for your `TestItem` class

https://github.com/MvvmCross/MvvmCross/pull/135

**Comment by slodge** [comment]
Actually you'll need to do more than that - not sure how you expect a `string` to equal a `TestItem`?

**Comment by JimWilcox3** [comment]
@slodge I didn't expect a string to equal a test item.  The problem is in the spinner which sets one field of the Testitem, "TagColor", which is a string.  I was loading the spinner's ItemsSource with an array of strings "TagColors".  I put the string[] in the viewmodel, because it was the only way I could find to be able to bind the ItemsSource on a spinner which resides within a ListView.   Please take another look at this.  I am new to GitHub but I will make a repo if you will look at it.

**Comment by JimWilcox3** [comment]
I just looked at it again and my Xml didn't show up properly.  Please take another look at this considering the proper Xml.  The string in the spinner should equal a string in the TestItem "TagColor".
Please take another look at this and consider the corrected Xml.  I am new to GitHub and wasn't expecting the format to change like that.  I will redo my original post using code fencing so you can see it better.

The xml for the View "TestView"  is:
```
<?xml version="1.0" encoding="utf-8"?>
<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
              xmlns:local="http://schemas.android.com/apk/res-auto"
              android:orientation="vertical"
              android:layout_width="fill_parent"
              android:layout_height="fill_parent">

  <Mvx.MvxListView
    android:layout_width="fill_parent"
    android:layout_height="wrap_content"
    android:layout_marginTop="5dp"
    local:MvxBind="ItemsSource Items"
    local:MvxItemTemplate="@layout/testitemview" />

</LinearLayout>
```

The layout for the MvxListView "TestItemView" is:
```
<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
              xmlns:local="http://schemas.android.com/apk/res-auto"
              android:orientation="horizontal"
              android:layout_width="fill_parent"
              android:layout_height="fill_parent">

  <MvxSpinner
    android:layout_width="130dp"
    android:layout_height="wrap_content"
    local:MvxBind="ItemsSource TagColors; SelectedItem TagColor"
    android:id="@+id/spinner1" />

  <EditText
    android:layout_width="130sp"
    android:layout_height="wrap_content"
    local:MvxBind="Text TagLot" />

  <EditText
    android:layout_width="150sp"
    android:layout_height="wrap_content"
    local:MvxBind="Text InventoryNumber" />
</LinearLayout>
```

Please look into this.  I really need it to work.  It is mission critical for me.

**Comment by slodge** [comment]
Sure I'll reopen it... back to where @Cheesebaron said "Could you make a GitHub repo with this code?"

**Comment by JimWilcox3** [comment]
@slodge Thanks.  I am working on the repo now.

**Comment by JimWilcox3** [comment]
@slodge Here is the repo you requested.  Thanks again for checking into this

https://github.com/JimWilcox3/MvxSpinnerTest

**Comment by Cheesebaron** [comment]
The problem lies in how Android works and not MvvmCross.

The problem is that you do two things wrong forcing Android to make a huge amount of redraw calls and calculations on how big views are.

First thing you do wrong is that you tell the `MvxListView` to `wrap_content`. Change this to `match_parent` or a fixed height.

Second thing you tell the row views to `wrap_content` this bogs down performance on `ListView`s massively because of the huge amount of measuring.

`wrap_content` is the root of all evil when dealing with `ListView`, basically... *never* use it.

**Comment by JimWilcox3** [comment]
@Cheesebaron Thanks for your help in this.  I did what you said and it fixed it in the test I posted.  I am still having the problem in my actual app.  Do I need to do the same thing with the LinearLayouts that contain the MvxListviews?.  Can you give me any good rule of thumb for laying things out?  My screen is pretty complicated. It has several nested LinearLayouts.  I originally wrote it in xaml for Windows RT where it was no problem to nest StackPanels and Grids.  I am having to port it to Android.  Android seems to take a significant performance hit when you go nesting things.

**Comment by JimWilcox3** [comment]
@slodge @Cheesebaron 
FYI I did a lot of testing since you told me about the wrap_content.  If you use it anywhere up the hierarchy from the MvxListView, it causes the problem I am having.  To be honest, this is quite inconvenient for me as I need to be able to size the screen according to how many items are on the list.  This worked well on Windows RT.  I guess Android is just not up to it.  In any case, unless there is something you can do to alleviate the issue with it having to issue multiple redraws to determine the height, then I guess this is not an MvvmCross problem.  Thanks so much for looking at it.

**Comment by Cheesebaron** [comment]
Usually what you achieve with nested `LinearLayouts` you can do a lot flatter with `RelativeLayout` it takes some effort to position things nicely and relatively to each other. The end result is so much better and requires less draw cycles. If you have some specific layouts you want me to look at make some gists or whatever where I can help you. Hit me up on Twitter `@Cheesebaron`

**Comment by JimWilcox3** [comment]
@Cheesebaron 
Here is my markup if you want to take a stab at it.  I put in fixed sizes everywhere that might be in the hierarchy of the two MvxItemLists.  This has fixed the immediate problem.  After testing this a lot this weekend, I think I have found a pattern.  If the MvxItemList has to automatically calculate its size, it has to render it once to see what that size will be.  I think that when this happens, it renders it with null values.  The "nulls not allowed in Selected Item" went away.  Perhaps adjustments cold be made to MvxSpinner and MvxAutoComplete so that when the value is null, it assumes that it is being used in an MvxItemTemplate as part of an MvxItemList.  This is just a thought.  Anyway here is my markup that works.  See what you think about this:

```
<?xml version="1.0" encoding="utf-8"?>
<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:local="http://schemas.android.com/apk/res-auto"
    xmlns:shared="MPS_Mobile_Shared.Common,assembly=MPS_Mobile_Driver.Droid"
    android:orientation="vertical"
    android:layout_width="fill_parent"
    android:layout_height="fill_parent">
    <LinearLayout
        android:orientation="horizontal"
        android:layout_width="fill_parent"
        android:layout_height="285dp"
        android:background="@drawable/doublewidthborder">
        <LinearLayout
            android:orientation="vertical"
            android:layout_width="0dp"
            android:layout_height="fill_parent"
            android:layout_weight=".7">
            <LinearLayout
                android:orientation="horizontal"
                android:layout_width="fill_parent"
                android:layout_height="wrap_content">
                <LinearLayout
                    android:layout_width="0dp"
                    android:layout_height="wrap_content"
                    android:orientation="vertical"
                    android:layout_weight="1">
                    <TextView
                        android:layout_width="fill_parent"
                        android:layout_height="wrap_content"
                        style="@style/InputTextView"
                        android:text="Date" />
                    <EditText
                        android:layout_width="fill_parent"
                        android:layout_height="wrap_content"
                        style="@style/InputEditText"
                        local:MvxBind="FocusText ShipmentInventory.InventoryDate, Converter=DateToStringConverter" />
                </LinearLayout>
                <LinearLayout
                    android:layout_width="0dp"
                    android:layout_height="wrap_content"
                    android:orientation="vertical"
                    android:layout_weight="1">
                    <TextView
                        android:layout_width="fill_parent"
                        android:layout_height="wrap_content"
                        style="@style/InputTextView"
                        android:text="Vehicle" />
                    <EditText
                        android:layout_width="fill_parent"
                        android:layout_height="wrap_content"
                        android:id="@+id/txtfleet"
                        style="@style/InputEditText"
                        local:MvxBind="Text ShipmentInventory.FleetNo" />
                </LinearLayout>
                <LinearLayout
                    android:layout_width="0dp"
                    android:layout_height="wrap_content"
                    android:orientation="vertical"
                    android:layout_weight="2">
                    <TextView
                        android:layout_width="fill_parent"
                        android:layout_height="wrap_content"
                        style="@style/InputTextView"
                        android:text="Agent" />
                    <EditText
                        android:layout_width="fill_parent"
                        android:layout_height="wrap_content"
                        android:id="@+id/txtagent"
                        style="@style/InputEditText"
                        local:MvxBind="Text ShipmentInventory.Agent" />
                </LinearLayout>
            </LinearLayout>
            <LinearLayout
                android:orientation="horizontal"
                android:layout_width="fill_parent"
                android:layout_height="wrap_content">
                <LinearLayout
                    android:layout_width="0dp"
                    android:layout_height="wrap_content"
                    android:orientation="vertical"
                    android:layout_weight="1">
                    <TextView
                        android:layout_width="fill_parent"
                        android:layout_height="wrap_content"
                        style="@style/InputTextView"
                        android:text="Shipment Number" />
                    <EditText
                        android:layout_width="fill_parent"
                        android:layout_height="wrap_content"
                        android:id="@+id/txtagent"
                        style="@style/InputEditText"
                        local:MvxBind="Text ShipmentInventory.idno" />
                </LinearLayout>
                <LinearLayout
                    android:layout_width="0dp"
                    android:layout_height="wrap_content"
                    android:orientation="vertical"
                    android:layout_weight="1">
                    <TextView
                        android:layout_width="fill_parent"
                        android:layout_height="wrap_content"
                        style="@style/InputTextView"
                        android:text="GBL Number" />
                    <EditText
                        android:layout_width="fill_parent"
                        android:layout_height="wrap_content"
                        android:id="@+id/txtagent"
                        style="@style/InputEditText"
                        local:MvxBind="Text ShipmentInventory.GBL" />
                </LinearLayout>
                <LinearLayout
                    android:layout_width="0dp"
                    android:layout_height="wrap_content"
                    android:orientation="vertical"
                    android:layout_weight="1">
                    <TextView
                        android:layout_width="fill_parent"
                        android:layout_height="wrap_content"
                        style="@style/InputTextView"
                        android:text="Shipper" />
                    <EditText
                        android:layout_width="fill_parent"
                        android:layout_height="wrap_content"
                        android:id="@+id/txtagent"
                        style="@style/InputEditText"
                        local:MvxBind="Text ShipmentInventory.Shipper" />
                </LinearLayout>
            </LinearLayout>
            <LinearLayout
                android:layout_width="fill_parent"
                android:layout_height="wrap_content"
                android:orientation="vertical">
                <TextView
                    android:layout_width="fill_parent"
                    android:layout_height="wrap_content"
                    style="@style/InputTextView"
                    android:text="Origin Load Address" />
                <EditText
                    android:layout_width="fill_parent"
                    android:layout_height="wrap_content"
                    style="@style/InputEditText"
                    local:MvxBind="Text ShipmentInventory.OrgAddress" />
            </LinearLayout>
            <LinearLayout
                android:layout_width="fill_parent"
                android:layout_height="wrap_content"
                android:orientation="vertical">
                <TextView
                    android:layout_width="fill_parent"
                    android:layout_height="wrap_content"
                    style="@style/InputTextView"
                    android:text="Destination Address" />
                <EditText
                    android:layout_width="fill_parent"
                    android:layout_height="wrap_content"
                    style="@style/InputEditText"
                    local:MvxBind="Text ShipmentInventory.DestAddress" />
            </LinearLayout>
        </LinearLayout>
        <LinearLayout
            android:orientation="vertical"
            android:layout_width="0dp"
            android:layout_height="fill_parent"
            android:layout_weight=".3"
            android:layout_marginLeft="5dp">
            <LinearLayout
                android:orientation="horizontal"
                android:layout_width="fill_parent"
                android:layout_height="65dp">
                <TextView
                    android:layout_width="140dp"
                    android:layout_height="fill_parent"
                    style="@style/TableHeaderTextView"
                    android:text="Tag Color"
                    android:background="@drawable/singlewidthborder" />
                <TextView
                    android:layout_width="160dp"
                    android:layout_height="fill_parent"
                    style="@style/TableHeaderTextView"
                    android:text="Inventory Lot Number"
                    android:background="@drawable/singlewidthborder" />
                <TextView
                    android:layout_width="60dp"
                    android:layout_height="fill_parent"
                    style="@style/TableHeaderTextView"
                    android:text=""
                    android:background="@drawable/singlewidthborder" />
            </LinearLayout>
            <LinearLayout
                android:layout_width="fill_parent"
                android:layout_height="139dp"
                android:orientation="vertical">
                <MvxListView
                    android:layout_width="fill_parent"
                    android:layout_height="95dp"
                    android:layout_marginTop="5dp"
                    local:MvxBind="ItemsSource ShipmentLots.Lots"
                    local:MvxItemTemplate="@layout/inventorylotview" />
                <ImageButton
                    android:src="@drawable/ic_action_new"
                    android:scaleType="fitCenter"
                    android:padding="5dp"
                    android:layout_width="40dp"
                    android:layout_height="40dp"
                    local:MvxBind="Click NewLot_Clicked"
                    android:id="@+id/btnLotNew" />
            </LinearLayout>
            <LinearLayout
                android:layout_width="fill_parent"
                android:layout_height="wrap_content"
                android:orientation="vertical">
                <TextView
                    android:layout_width="fill_parent"
                    android:layout_height="wrap_content"
                    style="@style/InputTextView"
                    android:text="Customer Login Name" />
                <EditText
                    android:layout_width="fill_parent"
                    android:layout_height="wrap_content"
                    android:id="@+id/txtCustomerLogin"
                    style="@style/InputEditText" />
            </LinearLayout>
        </LinearLayout>
    </LinearLayout>
  <ScrollView
    android:layout_width="fill_parent"
    android:layout_height="match_parent"
    android:layout_marginLeft="10dp">
    <HorizontalScrollView
      android:layout_width="fill_parent"
      android:layout_height="match_parent">
      <LinearLayout
        android:layout_width="fill_parent"
        android:layout_height="fill_parent"
        android:orientation="vertical">
        <Mvx.MvxListView
          android:layout_width="1690dp"
          android:layout_height="300dp"
          android:layout_marginTop="5dp"
          local:MvxBind="ItemsSource ShipmentInventory.Items"
          local:MvxItemTemplate="@layout/inventoryitemview" />
        <ImageButton
          android:src="@drawable/ic_action_new"
          android:scaleType="fitCenter"
          android:padding="5dp"
          android:layout_width="40dp"
          android:layout_height="40dp"
          local:MvxBind="Click NewItem_Clicked"
          android:id="@+id/btnItemNew" />
      </LinearLayout>
    </HorizontalScrollView>
  </ScrollView>
    <EditText
        android:inputType="date"
        android:layout_width="fill_parent"
        android:layout_height="wrap_content"
        android:text="Test Text"
        android:id="@+id/editText1" />
</LinearLayout>
```

**Comment by slodge** [comment]
I do hope you guys are also communicating via paypal

**Comment by JimWilcox3** [comment]
@slodge This conversation thread has helped me greatly.  It has also demonstrated a weakness in MvvmCross as using android:height="wrap_content", while inefficient, should be a legitimate way to do things.  All that being said, I decided days ago that if I am able to get my product to completion using MvvmCross, I plan to make a donation to you guys.  If my company won't do it, I will make a personal donation myself.  I am a hungry programmer too and I appreciate your product and the help you have given.

**Comment by JimWilcox3** [comment]
I redid my comment to @Cheesebaron since having my markup in there was outside the scope of the MvvmCross project.

**Comment by Cheesebaron** [comment]
@JimWilcox3 your observation is probably correct, although cases where the value is legitimately null when not in a measuring phase might have some implications. Not sure which though.

As for using `wrap_content`, why is there a requirement to use it compared to a fixed height? Is it because the views are having different heights depending on content, maybe something else?

**Comment by JimWilcox3** [comment]
@Cheesebaron I work for a moving company and am re-writing an app so our drivers can do an electronic inventory.  I am rewriting it for Android, because Microsoft has indicated that they will not be loading Windows RT/Phone on anything with a screen larger than 7".  It is also questionable as to whether they will have an alternative to the Nokia Lumia 2520 that we were using.  My boss decided we need to use the Samsung Tab4 10.1.  He wants it ported over by April O.o   The thing that I was able to do in the original RT version was to have the "add new line" button right after the last item on the list.  That way the drivers just tap the button, fill in that line, then tap the button again to create the next line.  You need a dynamic ListView height in order to make that work.  That's what I was trying to accomplish with the wrap_content.  It's not a deal breaker, but it was a nice feature.  I may try to see if I can capture a double tap on the blank area at the bottom of the list to add a new line.  Still bouncing around ideas on that one.

**Comment by Cheesebaron** [comment]
> was to have the "add new line" button right after the last item on the list. That way the drivers just tap the button, fill in that line, then tap the button again to create the next line. You need a dynamic ListView height in order to make that work.

That is not necessarily true.

If you have the `ListView` nested inside a `RelativeLayout` and have it set to `match_parent` *and* tell it to be above another view og viewgroup, which gives you a way to achieve that requirement.

**Comment by JimWilcox3** [comment]
@Cheesebaron Hey bro.  I never got it to work using a RelativeLayout.  When I did, it started doing the original problem again.  I have a pretty complicated layout so it could have probably been fixed, but with a lot of work.  I did get it to work using this though:

http://blog.masterdevs.com/headers-and-footers-on-an-mvxlistview/

I put the "add" button (among other things) in the footer.  Worked like a charm.
